disable plugins/tests which need rmw_fastrtps_cpp if unavailable

ivanpauno opened the pull request:
Fixes https://github.com/ros2/rosbag2/issues/136.

ivanpauno commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7427)](http://ci.ros2.org/job/ci_linux/7427/)

ivanpauno commented:
I think this is ready to merge

Karsten1987 commented:
I don't think that's the right approach.
We dynamically, but hardcoded, load `rmw_fastrtps_cpp` inside the code: https://github.com/ros2/rosbag2/blob/master/rosbag2_converter_default_plugins/src/rosbag2_converter_default_plugins/cdr/cdr_converter.cpp#L63

---
I believe we should always find `rmw_fastrtps_cpp` and don't compile the library at all if fastrtps is not available. With this patch, we currently build the library correctly, but it fails on runtime due to the missing libraries. That also explains the set of test failures (the segfaults) of the CI job you've linked.
When fast-rtps is not available, we should further disable all tests which rely on the converter plugins, as they are the only source requiring fastrtps.

ivanpauno commented:
> I don't think that's the right approach.
> We dynamically, but hardcoded, load `rmw_fastrtps_cpp` inside the code: https://github.com/ros2/rosbag2/blob/master/rosbag2_converter_default_plugins/src/rosbag2_converter_default_plugins/cdr/cdr_converter.cpp#L63
> 
> I believe we should always find `rmw_fastrtps_cpp` and don't compile the library at all if fastrtps is not available. With this patch, we currently build the library correctly, but it fails on runtime due to the missing libraries. That also explains the set of test failures (the segfaults) of the CI job you've linked.
> When fast-rtps is not available, we should further disable all tests which rely on the converter plugins, as they are the only source requiring fastrtps.

Ok, I will check what has to be disabled an update the PR.

ivanpauno commented:
@Karsten1987 In https://github.com/ros2/rosbag2/pull/137/commits/bb38b7a1b65771b9cf14f51bcb1590203bac36b1, I completely disabled the build of `rosbag2_converter_default_plugins` if you don't have `fastrtps`. I also disabled the three tests in `rosbag2_tests` which I think that they depend at runtime in the former package.

I also saw in the previous CI that `test_rosbag2_node` and `test_play` were failing, but I'm not sure if they also depend on the converter plugin, or if it was a random error related to using `opensplice`.
Can you check please?
Thanks!

ivanpauno commented:
@Karsten1987 Friendly ping

Karsten1987 commented:
Sorry @ivanpauno I am currently pretty underwater and won't have time to deep dive into it for the next two weeks. 

It makes sense to me to completely disable the converter plugin tests when fast rtps is not available. I could further imagine (without having precisely looked at the code) that some end-to-end tests might have to be disabled or compiled differently when the converters are not present in the current setup.

ivanpauno commented:
> Sorry @ivanpauno I am currently pretty underwater and won't have time to deep dive into it for the next two weeks.
> 
> It makes sense to me to completely disable the converter plugin tests when fast rtps is not available. I could further imagine (without having precisely looked at the code) that some end-to-end tests might have to be disabled or compiled differently when the converters are not present in the current setup.

No problem, I will wait then. Thanks for the update!

Karsten1987 commented on rosbag2_converter_default_plugins/package.xml:
```diff
@@ -14,8 +14,10 @@
+  <depend>rmw_fastrtps_cpp</depend>
   <depend>rosbag2</depend>
 
+
```
superfluous new line

dirk-thomas replied:
Please revert unrelated (whitespace) changes.

Karsten1987 commented on rosbag2_converter_default_plugins/CMakeLists.txt:
```diff
@@ -24,55 +24,57 @@ find_package(rosbag2 REQUIRED)
 
-add_library(${PROJECT_NAME} SHARED
-  src/rosbag2_converter_default_plugins/cdr/cdr_converter.cpp)
+if(rmw_fastrtps_cpp_FOUND)
```
I would somehow inverse this logic and return `if(NOT rmw_fastrtps_cpp_FOUND)` with at least a notification that this library is not going to be compiled in full. This will also ease the diff a little bit as we don't have an extra layer of indentation.

dirk-thomas commented:
Please run CI for this change with a non-FastRTPS RMW impl. If the build doesn't fail (which would be a big step forward) we will get this merged in the current state.

dirk-thomas commented:
> I believe the right change here is to make the plugins work with whatever RMW implementation available and try to avoid a hardcoded version of rmw_fastrtps_cpp.

That might be a good goal but is outside of the scope of this PR.

ivanpauno commented:
I've addressed comments and rebased with master.

CI (only opensplice, up to rosbag2 related packages):
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7747)](http://ci.ros2.org/job/ci_linux/7747/)

ivanpauno commented:
Again:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7753)](http://ci.ros2.org/job/ci_linux/7753/)

ivanpauno commented:
> Again:
> 
> * Linux [![Build Status](https://camo.githubusercontent.com/1f59b4a3eedfa5015563462cb6eba4d2ccc29c3d/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e7578266275696c643d37373533)](http://ci.ros2.org/job/ci_linux/7753/)

It builds correctly now (only opensplice). I solved the linter failures.
There are two other build failures, but they shouldn't happen if fastrtps is available (they don't appear locally, I have the three rmw implementations).

Double checking (fastrtps and connext, same packages as before):
* Linux [![Build Status]* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7756)](http://ci.ros2.org/job/ci_linux/7756/)

ivanpauno commented:
> Double checking (fastrtps and connext, same packages as before):
> 
> * Linux [![Build Status](https://camo.githubusercontent.com/ca867ca7bcd3c55fd46c27fcbac89432150ac682/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e7578266275696c643d37373536)](http://ci.ros2.org/job/ci_linux/7756/)

Again, because I chose opensplice and connext by error :man_facepalming: :
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7758)](http://ci.ros2.org/job/ci_linux/7758/)

Karsten1987 commented on rosbag2_transport/package.xml:
```diff
@@ -14,6 +14,7 @@
   <depend>rmw</depend>
   <depend>shared_queues_vendor</depend>
 
+  <test_depend>ament_index_cpp</test_depend>
```
nitpick: Alphabetical order

dirk-thomas commented:
@Karsten1987 can you please make sure that this gets backported into Dashing in time for the upcoming patch release.

dirk-thomas commented:
> can you please make sure that this gets backported into Dashing in time for the upcoming patch release.

Once the fix has been merged to the `dashing` branch the following job should start to pass (or at least become unstable): http://build.ros2.org/view/Dci/job/Dci__nightly-cyclonedds_ubuntu_bionic_amd64/
